service/remotecfg/remotecfg.go

```diff
@@ -118,12 +133,93 @@ var _ service.Service = (*Service)(nil)
 func (s *Service) Run(ctx context.Context, host service.Host) error {
 	s.ctrl = host.NewController(ServiceName)
 
+	// Run the service's own controller.
 	go func() {
 		s.ctrl.Run(ctx)
 	}()
+
+	// We're on the initial start-up of the service.
+	// Let's try to read from the API, parse and load the response contents.
+	// If either the request itself or parsing its contents fails, try to read
+	// from the on-disk cache on a best-effort basis.
+	var (
+		gcr              *connect.Response[agentv1.GetConfigResponse]
+		src              *flow.Source
+		getErr, parseErr error
+	)
+	gcr, getErr = s.asClient.GetConfig(ctx, s.getConfigRequest)
+	if getErr == nil {
+		src, parseErr = flow.ParseSource(ServiceName, []byte(gcr.Msg.GetContent()))
+	}
+
+	// Reading from the API succeeded, let's try to load the contents.
+	if getErr == nil && parseErr == nil {
+		err := s.ctrl.LoadSource(src, nil)
+		if err != nil {
+			level.Error(s.opts.Logger).Log("msg", "could not load the API response contents", "err", err)
+		} else {
+			s.currentConfigHash = getHash(gcr.Msg.Content)
+		}
+	} else {
+		// Either the API call or parsing its contents failed, let's try the
+		// on-disk cache.
+		level.Info(s.opts.Logger).Log("msg", "falling back to the on-disk cache")
+		b, err := os.ReadFile(s.dataPath)
+		if err != nil {
+			level.Error(s.opts.Logger).Log("msg", "could not read from the on-disk cache", "err", err)
+		}
+		if len(b) > 0 {
+			src, err := flow.ParseSource(ServiceName, b)
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "could not parse the on-disk cache contents", "err", err)
+			} else {
+				err = s.ctrl.LoadSource(src, nil)
+				if err != nil {
+					level.Error(s.opts.Logger).Log("msg", "could not load the on-disk cache contents", "err", err)
+				} else {
+					s.currentConfigHash = getHash(string(b))
+				}
+			}
+		}
+	}
+
 	for {
 		select {
+		case <-s.ticker.C:
```

Comment:
The code here feels a bit repetitive, and it looks like it could be made more DRY. I'd like for the Run function to be refactored a bit to carve out reusable bits of logic.